olive/model/hf_utils.py

```diff
@@ -25,6 +27,137 @@ class HFComponent(ConfigBase):
+    torch_dtype: str = None
+    # A map that specifies where each submodule should go. Refer to link in class docstring
+    # str suffices for torch.dtype, otherwise serializer won't recognize it
+    device_map: Union[int, str, Dict] = None
+    # A dictionary device identifier to maximum memory
+    max_memory: Dict = None
+    # quant method to use (transformers.QuantizationMethod), e.g. "bitsandbytes", "gptq"
+    # must be provided if quantization_config is provided
+    quantization_method: str = None
+    # A dictionary of configuration parameters for quantization
+    quantization_config: Dict = None
+    # other kwargs
+    extra_args: Dict = None
+
+    @validator("torch_dtype", pre=True)
+    def validate_torch_dtype(cls, v):
+        if isinstance(v, torch.dtype):
+            v = str(v).replace("torch.", "")
+        return v
+
+    @validator("device_map", pre=True)
+    def validate_device_map(cls, v):
+        if isinstance(v, torch.device):
+            v = cls.device_to_str(v)
+        elif isinstance(v, dict):
+            v = {k: cls.device_to_str(v) for k, v in v.items()}
+        return v
+
+    @validator("quantization_config", pre=True, always=True)
+    def validate_quantization_config(cls, v, values):
+        if "quantization_method" not in values:
+            # to ensure we don't get a KeyError
+            raise ValueError("Invalid quantization_config")
+        if (values["quantization_method"] and not v) or (not values["quantization_method"] and v):
+            raise ValueError("quantization_config and quantization_method must be provided together")
+        if not v:
+            return v
+
+        try:
+            return cls.dict_to_quantization_config(values["quantization_method"], v).to_dict()
+        except ImportError:
+            # we don't want to fail since the pass target might have the correct transformers version
+            logger.warning(
+                f"Could not import the config class for quantization method {values['quantization_method']}. Skipping "
+                " validation"
+            )
+            return v
+
+    def get_loading_args(self):
+        loading_args = {}
+        # copy args that can be directly copied
+        direct_copy_args = ["device_map", "max_memory"]
+        for arg in direct_copy_args:
+            if getattr(self, arg):
+                loading_args[arg] = deepcopy(getattr(self, arg))
+        # convert torch dtype to torch.dtype or "auto"
+        if self.torch_dtype:
+            loading_args["torch_dtype"] = self.get_torch_dtype()
+        # convert quantization_config to the config class
+        quantization_config = self.get_quantization_config()
+        if quantization_config:
+            loading_args["quantization_config"] = quantization_config
+        # add extra args
+        if self.extra_args:
+            loading_args.update(deepcopy(self.extra_args))
+        return loading_args
+
+    def get_torch_dtype(self):
```

Comment:
>get_torch_dtype [](http://example.com/codeflow?start=8&length=15)

could we merge the logic here with the validator?

Reply:
the validator uses this helper function twice so I abstracted it. But I can make it a function inside the validator if you think it's better to not expose this method.

Reply:
This is a similar case to the `get_quantization_config` method. `torch.dtype` is not directly serializable but `from_pretrained` needs the value as torch.dtype.

Reply:
Similar comment to above